erigon-lib/commitment/commitment.go

```diff
@@ -758,3 +758,230 @@ func DecodeBranchAndCollectStat(key, branch []byte, tv TrieVariant) *BranchStat
 	}
 	return stat
 }
+
+// Defines how to evaluate commitments
+type Mode uint
+
+const (
+	ModeDisabled Mode = 0
+	ModeDirect   Mode = 1
+	ModeUpdate   Mode = 2
+)
+
+func (m Mode) String() string {
+	switch m {
+	case ModeDisabled:
+		return "disabled"
+	case ModeDirect:
+		return "direct"
+	case ModeUpdate:
+		return "update"
+	default:
+		return "unknown"
+	}
+}
+
+func ParseCommitmentMode(s string) Mode {
+	var mode Mode
+	switch s {
+	case "off":
+		mode = ModeDisabled
+	case "update":
+		mode = ModeUpdate
+	default:
+		mode = ModeDirect
+	}
+	return mode
+}
+
+type UpdateTree struct {
+	tree   *btree.BTreeG[*KeyUpdate]
+	keccak cryptozerocopy.KeccakState
+	keys   map[string]struct{}
+	mode   Mode
+}
+
+func NewUpdateTree(m Mode) *UpdateTree {
+	return &UpdateTree{
+		tree:   btree.NewG[*KeyUpdate](64, keyUpdateLessFn),
+		keccak: sha3.NewLegacyKeccak256().(cryptozerocopy.KeccakState),
+		keys:   map[string]struct{}{},
+		mode:   m,
+	}
+}
+
+func (t *UpdateTree) get(key []byte) (*KeyUpdate, bool) {
+	c := &KeyUpdate{plainKey: key, update: Update{CodeHashOrStorage: EmptyCodeHashArray}}
+	el, ok := t.tree.Get(c)
+	if ok {
+		return el, true
+	}
+	c.plainKey = common.Copy(c.plainKey)
+	return c, false
+}
+
+// TouchPlainKey marks plainKey as updated and applies different fn for different key types
+// (different behaviour for Code, Account and Storage key modifications).
+func (t *UpdateTree) TouchPlainKey(key string, val []byte, fn func(c *KeyUpdate, val []byte)) {
+	switch t.mode {
+	case ModeUpdate:
+		item, _ := t.get([]byte(key))
```

Comment:
1. `t.get` may accept string. 
2. `t.get` + `t.replace` seems bad pattern - because `t.get` returns pointer to existing in tree `item` (can make problem for "parallel commitment" implementation, or maybe somebody will use `.get` and modify result). I think need replace it by 1 func - which doesn't return anything.